payments/tests/test_bambora_payform_refund.py

```diff
@@ -69,6 +69,58 @@ def test_initiate_refund_success(provider_base_config: dict, order: Order):
     )
 
 
+@pytest.mark.parametrize(
+    "order", ["berth_order", "winter_storage_order"], indirect=True,
+)
+def test_initiate_refund_no_order_email(provider_base_config: dict, order: Order):
+    """Test the request creator constructs the payload base and returns a url that contains a token"""
+    request = RequestFactory().request()
+    order.status = OrderStatus.PAID
+    order.email = None
+    order.lease.status = LeaseStatus.PAID
+    order.lease.save()
+    order.save()
+
+    OrderToken.objects.create(
+        order=order, token="98765", valid_until=now() - relativedelta(hours=1)
+    )
+
+    payment_provider = create_bambora_provider(provider_base_config, request)
+    with mock.patch(
+        "payments.providers.bambora_payform.requests.post",
+        side_effect=mocked_refund_response_create,
+    ):
+        refund = payment_provider.initiate_refund(order)
+
+    assert refund.refund_id == "123456"
+
+
+@pytest.mark.parametrize(
+    "order", ["berth_order", "winter_storage_order"], indirect=True,
+)
+def test_initiate_refund_no_order_email_or_application(
+    provider_base_config: dict, order: Order
+):
+    """Test the request creator constructs the payload base and returns a url that contains a token"""
+    request = RequestFactory().request()
+    order.status = OrderStatus.PAID
+    order.customer_email = None
+    order.lease.status = LeaseStatus.PAID
+    order.lease.application = None
+    order.lease.save()
+    order.save()
+
+    OrderToken.objects.create(
+        order=order, token="98765", valid_until=now() - relativedelta(hours=1)
```

Comment:
Here too, token valid_until is in the past, doesn't change the result though

Reply:
Ah yep those slipped in there :sweat_smile: Good spot!